### PR TITLE
feat: Add support for Auto Scaling Group Instance Refresh for self-managed worker groups

### LINKED ISSUE #929: Instance Refresh for EC2 Auto Scaling
## I'm submitting a...

* [ ] bug report
* [x] feature request
* [ ] support request - read the [FAQ](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/faq.md) first!
* [ ] kudos, thank you, warm fuzzy

## What is the current behavior?

When this module makes a change to the worker group launch template an external action is required to update the workers to use the new launch template.

## If this is a bug, how to reproduce? Please include a code sample if relevant.

n/a

## What's the expected behavior?

I'm interested to see if the new [Instance Refresh for EC2 Auto Scaling](https://aws.amazon.com/blogs/compute/introducing-instance-refresh-for-ec2-auto-scaling/) would be something that could be leveraged by this module to automate worker upgrades?

Currently the best solution I've found for this is the [HelloFresh EKS Rolling Update](https://github.com/hellofresh/eks-rolling-update) but due to the fact that this is synchronous and time consuming it doesn't seem like a good fit for a _Terraform_ module. I'd be interested to see if _Instance Refresh for EC2 Auto Scaling_ is an alternative that can be embedded in a _Terraform_ module.

## Are you able to fix this problem and submit a PR? Link here if you have already.

n/a

## Environment details

n/a

## Any other relevant info

n/a

### PR DESCRIPTION
**Pull request by bashims**
# PR o'clock

## Description

Add support for the instance_refresh feature of the AWS provisioner added in https://github.com/hashicorp/terraform-provider-aws/pull/16678

This PR resolves #929 

See:
https://aws.amazon.com/blogs/compute/introducing-instance-refresh-for-ec2-auto-scaling/

### Checklist

- [x] CI tests are passing
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by srinirei** [comment]
Hi, I am using eks module in my project and over the weekend I started to see the following issue

``Error: AuthFailure: AWS was not able to validate the provided access credentials
        status code: 401, request id: 2b815ee1-e91a-4878-8a31-32f8b72fd2a8
  on .terraform/modules/eks/data.tf line 16, in data "aws_ami" "eks_worker":
  16: data "aws_ami" "eks_worker" {

Error: AuthFailure: AWS was not able to validate the provided access credentials
        status code: 401, request id: b6d1e8c6-f6ee-48ec-9e20-3cc7c59d63a3
  on .terraform/modules/eks/data.tf line 27, in data "aws_ami" "eks_worker_windows":
  27: data "aws_ami" "eks_worker_windows" {``

I am using the module in my code as follows

module "eks" {
  source = "terraform-aws-modules/eks/aws"
  version         = "12.1.0"
  cluster_name    = local.cluster_name
  cluster_version = local.cluster_version
  vpc_id          = module.vpc.aws_vpc_id
  subnets         = module.vpc.aws_subnet_private_prod_ids
  cluster_enabled_log_types = ["api", "audit", "authenticator", "controllerManager", "scheduler"]
  cluster_log_retention_in_days = 180
  manage_aws_auth = false
  write_kubeconfig = false
}

**Comment by stevehipwell** [comment]
@bashims does the [phased deployments anouncement](https://aws.amazon.com/about-aws/whats-new/2021/03/amazon-ec2-auto-scaling-instance-refresh-supports-phased-deployments/) make any difference to this PR?

It looks like you haven't configured the ASGs to work with the [aws-node-termination-handler](https://github.com/aws/aws-node-termination-handler) (queue processor mode, [chart](https://github.com/aws/eks-charts/tree/master/stable/aws-node-termination-handler))? I would have expected this to be a requirement as you wouldn't want a node to suddenly terminate with no warning.

**Comment by bashims** [comment]
> @bashims does the [phased deployments anouncement](https://aws.amazon.com/about-aws/whats-new/2021/03/amazon-ec2-auto-scaling-instance-refresh-supports-phased-deployments/) make any difference to this PR?

Good question! I am not too sure if the checkpoint feature is available in the AWS provider. For now it seems to support the original instance refresh feature.

> 
> It looks like you haven't configured the ASGs to work with the [aws-node-termination-handler](https://github.com/aws/aws-node-termination-handler) (queue processor mode, [chart](https://github.com/aws/eks-charts/tree/master/stable/aws-node-termination-handler))? I would have expected this to be a requirement as you wouldn't want a node to suddenly terminate with no warning.

Good point. I can add something to the docs to mention that one should deploy the `aws-node-termination-handler` chart when enabling instance refresh. The eks TF module does not provide support for deploying charts etc, so I doubt it would be a good fit here.

**Comment by stevehipwell** [comment]
@bashims I wasn't refering to actually installing the chart but you'll need to set up a SQS queue for the events and then provide the chart installation instructions to work with instance refresh (specifically controller not daemonset mode).

**Comment by bashims** [comment]
> @bashims I wasn't refering to actually installing the chart but you'll need to set up a SQS queue for the events and then provide the chart installation instructions to work with instance refresh (specifically controller not daemonset mode).

@stevehipwell  Awesome feedback. I was not aware. I will update this PR to include this feature.

**Comment by Erokos** [comment]
Hi, what's the status with this PR?

**Comment by Erokos** [comment]
Hi,
one other question regarding the work in this PR is, when this gets merges and used, will the "asg_recreate_on_change" param still be used/necessary?

**Comment by barryib** [comment]
> Hi,
> one other question regarding the work in this PR is, when this gets merges and used, will the "asg_recreate_on_change" param still be used/necessary?

Good question. I don't think so. I think with this addition we can drop the `asg_recreate_on_change` and random pet names.

@bashims @stevehipwell I think we should add in docs that using instance refresh will recreate your nodes, but it doesn't drains them. To do that, users should install https://github.com/aws/aws-node-termination-handler

Related to:
- https://github.com/terraform-aws-modules/terraform-aws-eks/issues/462

**Comment by stevehipwell** [comment]
@bashims @barryib I assume that this PR will add the SQS queue that _aws-node-termination-handler_ needs. So if instance refresh is turned on you need to install _aws-node-termination-handler_ for draining but everything else works?

**Comment by Erokos** [comment]
I wholeheartedly agree. From reading the issues, I've come across references on random_pet usage and from the docs it isn't highlighted that using `asg_recreate_on_change` doesn't drain nodes and in my opinion it should say so. To some it may be obvious but a warning never hurt nobody 🙂. I could be wrong but without having to manage the draining yourself, by having a shell script within a null resource (with a remote-exec) or something similar, I don't see how else it could be done.

**Comment by barryib** [comment]
> @bashims @barryib I assume that this PR will add the SQS queue that _aws-node-termination-handler_ needs. So if instance refresh is turned on you need to install _aws-node-termination-handler_ for draining but everything else works?

Humm... But users should handle aws-node-terminination-handler installation with all requirements themselves. In this module, we'll only support changes in worker ASG like:
- instance refresh configuration
- lifecycle hooks configuration

Everything else should be handled outside of this module. We can write docs for that. If someone willing to write a module to address aws-node-termination-handler installation, we can probably include that link in docs.

**Comment by stevehipwell** [comment]
@barryib I think the SQS linking is part of the ASG configuration, so at the minimum that would need to be passed into the module so the _aws-node-termination-handler_ can be linked up correctly.

**Comment by barryib** [comment]
> @barryib I think the SQS linking is part of the ASG configuration, so at the minimum that would need to be passed into the module so the _aws-node-termination-handler_ can be linked up correctly.

Yep. It'll be part of the lifecycle hook configuration https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/local.tf#L37

**Comment by bashims** [comment]
> > Hi,
> > one other question regarding the work in this PR is, when this gets merges and used, will the "asg_recreate_on_change" param still be used/necessary?
> 
> Good question. I don't think so. I think with this addition we can drop the `asg_recreate_on_change` and random pet names.
> 
> @bashims @stevehipwell I think we should add in docs that using instance refresh will recreate your nodes, but it doesn't drains them. To do that, users should install https://github.com/aws/aws-node-termination-handler
> 
> Related to:
> 
>     * #462

I can take care of that in this PR.

**Comment by barryib** [comment]
> I can take care of that in this PR.

Thanks @bashims. Can you please:
- Update docs to highlight the need of aws-node-termination handler for draining
- Open a new PR (if you have time) to drop `asg_recreate_on_change` and random_pets in for LT and LC

File: variables.tf
Comment (james-callahan) [inline variables.tf:166]: Rather than have this module-wide, could we make this a per-ASG setting?

File: workers.tf
Comment (james-callahan) [inline workers.tf:157]: e.g. via:
```suggestion
    for_each = lookup(
        var.worker_groups[count.index],
        "instance_refresh_enabled",
        local.workers_group_defaults["instance_refresh_enabled"]
      ) ? [1] : []
```

**Review by james-callahan — CHANGES_REQUESTED** [review]

File: variables.tf
Comment (barryib) [reply variables.tf:166]: Good point. Agreed.

File: workers.tf
Comment (Erokos) [reply workers.tf:157]: this also implies a default value set in locals.tf

File: variables.tf
Comment (bashims) [reply variables.tf:166]: Sounds good.

Comment (bashims) [reply variables.tf]: Dropped the global variable. One can set `instance_refresh_enabled` globally with:

``` 
workers_group_defaults = {
    "instance_refresh_enabled" : true
  }
```

File: workers.tf
Comment (bashims) [reply workers.tf]: Done in my most recent commit.

**Comment by bashims** [comment]
> > @bashims @barryib I assume that this PR will add the SQS queue that _aws-node-termination-handler_ needs. So if instance refresh is turned on you need to install _aws-node-termination-handler_ for draining but everything else works?
> 
> Humm... But users should handle aws-node-terminination-handler installation with all requirements themselves. In this module, we'll only support changes in worker ASG like:
> 
>     * instance refresh configuration
> 
>     * lifecycle hooks configuration
> 
> 
> Everything else should be handled outside of this module. We can write docs for that. If someone willing to write a module to address aws-node-termination-handler installation, we can probably include that link in docs.

**Comment by bashims** [comment]
@barryib I think this PR is nearly ready. I am working on complete example that includes necessary setup for graceful  node draining with instance refresh. I should have this done soon.

File: local.tf
Comment (james-callahan) [inline local.tf:46]: Why change this? It was recently changed in the opposite direction by https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1208

File: local.tf
Comment (james-callahan) [inline local.tf:4]: this is invalid...

**Review by james-callahan — CHANGES_REQUESTED** [review]

File: local.tf
Comment (bashims) [reply local.tf:46]: Good catch. This was due to my rebase off of main. Fixing.

File: local.tf
Comment (bashims) [reply local.tf:4]: ditto

**Review by james-callahan — APPROVED** [review]

File: local.tf
Comment (bashims) [reply local.tf]: reverted this change.

File: local.tf
Comment (bashims) [reply local.tf]: reverted

**Comment by bashims** [comment]
I think that this PR is ready to go. The complete example works perfectly.

Just waiting on workflow approval, and final review. Once that is done I will open a PR for dropping `asg_recreate_on_change` and random pet names.

File: examples/instance_refresh/main.tf
Comment (barryib) [inline examples/instance_refresh/main.tf:149]: Why can't we use the [helm terraform provider](https://registry.terraform.io/providers/hashicorp/helm/latest/docs) to install the node termination handler ?

File: examples/instance_refresh/main.tf
Comment (barryib) [inline examples/instance_refresh/main.tf:219]: If you create this lifecycle hook, you don't need to set worker_create_initial_lifecycle_hook to true.

File: examples/instance_refresh/main.tf
Comment (barryib) [inline examples/instance_refresh/main.tf:127]:
```suggestion
      "resources" : module.eks.workers_asg_arns
```

File: examples/instance_refresh/main.tf
Comment (barryib) [inline examples/instance_refresh/main.tf:217]: Why can't we use the [helm terraform provider](https://registry.terraform.io/providers/hashicorp/helm/latest/docs) to install the node termination handler ?

Comment (bashims) [reply examples/instance_refresh/main.tf:217]: We tried it in the past and had some issues with it. Perhaps those are issues no longer exist. I will give it a try.

Comment (barryib) [reply examples/instance_refresh/main.tf:217]: please let me know, I would like to merge this tonight or tomorrow.

Comment (audriusb) [reply examples/instance_refresh/main.tf:217]: Please to it tonight. Thanks :-)

Comment (stevehipwell) [reply examples/instance_refresh/main.tf:217]: I can confirm that the latest Helm provider can install the _aws-node-termination-handler_, we've had no issues with this at all.

Comment (bashims) [reply examples/instance_refresh/main.tf:217]: Yes, it's working fine now. I wanted to confirm that destroy ran cleanly.

Comment (bashims) [reply examples/instance_refresh/main.tf:217]: Pushing the requested changes soon.

File: examples/instance_refresh/main.tf
Comment (bashims) [reply examples/instance_refresh/main.tf:219]: fixed.

File: examples/instance_refresh/main.tf
Comment (bashims) [reply examples/instance_refresh/main.tf:127]: fixed.

File: examples/instance_refresh/main.tf
Comment (barryib) [reply examples/instance_refresh/main.tf:219]: @bashims did you read this https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/autoscaling_group ?

image

 I was wondering if it's not better to use initial_lifecycle_hooks https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/local.tf#L37.

@bashims @stevehipwell @antonbabenko any thoughts ?

Comment (antonbabenko) [reply examples/instance_refresh/main.tf:219]: I think @bryantbiggs knows better about this.

Comment (bashims) [reply examples/instance_refresh/main.tf:191]: I tried the `initial_lifecycle_hook` and there is an issue on destroy. Basically the ANTH cannot complete the lifecycle action since it is terminated before the ASG is destroyed. This causes Terraform to timeout waiting for the ASG's destruction.

Comment (barryib) [reply examples/instance_refresh/main.tf:191]: ohh thanks. Can you please add this as a comment in the example. It'll avoid asking the same question again and again ^^.

File: examples/instance_refresh/versions.tf
Comment (barryib) [inline examples/instance_refresh/versions.tf:8]: We don't need `null` and `template` provider anymore. Please remove them.

File: examples/instance_refresh/main.tf
Comment (bashims) [reply examples/instance_refresh/main.tf:191]: Absolutely.

**Comment by barryib** [comment]
@bashims you need to update your branch and resolve conflicts.

File: examples/instance_refresh/versions.tf
Comment (bashims) [reply examples/instance_refresh/versions.tf]: done

**Comment by bashims** [comment]
@barryib are you okay if I squash and rebase off of master?

**Comment by barryib** [comment]
> @barryib are you okay if I squash and rebase off of master?

Yep.

**Review by barryib — APPROVED** [review]

File: examples/instance_refresh/main.tf
Comment (barryib) [inline examples/instance_refresh/main.tf:26]:
```suggestion
provider "helm" {
  kubernetes {
    host                   = data.aws_eks_cluster.cluster.endpoint
    cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
    token                  = data.aws_eks_cluster_auth.cluster.token
  }
}
```

File: examples/instance_refresh/main.tf
Comment (barryib) [inline examples/instance_refresh/main.tf:193]:
```suggestion
# Creating the lifecycle-hook outside of the ASG resource's `initial_lifecycle_hook`
# ensures that node termination does not require the lifecycle action to be completed,
# and thus allows the ASG to be destroyed cleanly.
```

**Comment by barryib** [comment]
Thanks @bashims for your contributions.

**Comment by barryib** [comment]
This now shipped in [v16.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/releases/tag/v16.0.0)

**Pull request merged**
